### DISCUSSION TITLE
Mapping output affected by the order of base class maps in a profile

### DISCUSSION THREAD

**Discussion by danielgreen** [LuckyPennySoftware/AutoMapper#4094] [NA]
The order that maps are set up in a profile seems to affect the output of mapping operations.

Full repro in this minimal .NET 6 console app:
https://github.com/danielgreen/AutoMapperScratchApp

### Source/destination types

```csharp
// Source type and interface
public interface IFoo { public string? FooText { get; set; } }
public class Foo : IFoo { public string? FooText { get; set; } }

// Destination type and base class
public abstract class Bar { public string? FooText { get; set; } }
public class Boo : Bar { }
```

### Mapping configuration

```csharp
public class ProfileA : Profile
{
    public ProfileA()
    {
        this.CreateMap<object, IFoo>()
            .ForMember(d => d.FooText, o => o.Ignore())
            .IncludeAllDerived();

        this.CreateMap<Bar, IFoo>()
            .ForMember(d => d.FooText, o => o.MapFrom(s => s.FooText))
            .IncludeAllDerived();

        this.CreateMap<Boo, Foo>();
    }
}

public class ProfileB : Profile
{
    public ProfileB()
    {
        this.CreateMap<Bar, IFoo>()
            .ForMember(d => d.FooText, o => o.MapFrom(s => s.FooText))
            .IncludeAllDerived();

        this.CreateMap<object, IFoo>()
            .ForMember(d => d.FooText, o => o.Ignore())
            .IncludeAllDerived();

        this.CreateMap<Boo, Foo>();
    }
}
```

### Expected behavior

Expected the behaviour of both profiles would be the same.
Thought that FooText would be mapped in both cases, because the <Bar, IFoo> map (which maps the property) is more derived than the <object, IFoo> map (which ignores the property).

That was based on my reading of these docs:
https://docs.automapper.org/en/stable/Mapping-inheritance.html

### Actual behavior

The FooText property value is not mapped when using ProfileA, but the value is mapped when using ProfileB.
The only difference between ProfileA and ProfileB is the <Bar, IFoo> and <object, IFoo> maps appear in a different order.

Is it expected behaviour that the order that maps appear in a profile would determine the output?
If it is expected, is that documented (sorry if I missed it), or how should developers work with that?
If it is not expected, which profile is producing the correct output here?

### Steps to reproduce

```csharp
Console.WriteLine("AutoMapper scratch app");

TestProfileA();
TestProfileB();

Console.WriteLine("done");

void TestProfileA()
{
    var config = new MapperConfiguration(c => c.AddProfile<ProfileA>());
    config.AssertConfigurationIsValid();

    var mapper = config.CreateMapper();
    var boo = new Boo { FooText = "Hi" };
    var foo = mapper.Map<Foo>(boo);

    Console.WriteLine("Testing ProfileA - what is the value of Foo.FooText?");
    Console.WriteLine(foo.FooText ?? "(null)");
    Console.WriteLine();
}

void TestProfileB()
{
    var config = new MapperConfiguration(c => c.AddProfile<ProfileB>());
    config.AssertConfigurationIsValid();

    var mapper = config.CreateMapper();
    var boo = new Boo { FooText = "Hi" };
    var foo = mapper.Map<Foo>(boo);

    Console.WriteLine("Testing ProfileB - what is the value of Foo.FooText?");
    Console.WriteLine(foo.FooText ?? "(null)");
    Console.WriteLine();
}
```

**Comment by lbargaoanu** [comment]
Just by looking at your maps, I cannot say what the result should be :)
So just don't write this code. What happens if you simply use `Include` when you need it, without `IncludeAllDerived`?
Either way, you're welcome to look deeper into it and come up with a proposal.
But my point is that someone is supposed to look at your mappings and have a good idea what they mean. That seems very unlikely here.

**Reply by danielgreen** [reply]
I guess by simplifying the code into Foos and Bars I hoped it would be easier to inspect not more difficult :)

I just wanted to provide a sample that demonstrated the order of maps having an impact, when I didn't think that was expected.

In my real world project, the codebase uses the same inheritance pattern as the sample code, with interfaces and abstract classes used in the same way, which is how I stumbled on the behaviour.

Fundamentally, I have various destination classes that implement an interface with a property FooText.
Some of the source classes also have a FooText and I want that value to be mapped.
Some other source classes do not have it, so for them the destination FooText should be ignored (to satisfy AutoMapper validation).

That's why I define the <object, IFoo> and <Bar, IFoo> maps, to satisfy those two scenarios.

At your suggestion I've tried removing IncludeAllDerived, and instead used IncludeBase within the <Boo, Foo> map for the concrete types.
If I have one IncludeBase call to <object, IFoo>, the value is ignored, as expected for that base mapping.
If I have one IncludeBase call to <Bar, IFoo> the value is copied, as expected for that base mapping.

If I use both IncludeBase calls, then the order of them seems to determine the output.
If IncludeBase<object, IFoo> appears first then the value is ignored. If IncludeBase<Bar, IFoo> appears first then the value is copied.
So that's similar to the behaviour for IncludeAllDerived, except now it's the order of the IncludeBase calls that matters, not the order of the base maps in the profile.

I understand if that's not something that concerns you, I have maybe been unlucky that the way my real world codebase is structured caused me to trip over it. However I'm interested in your opinion about whether the order of maps or includes should or should not matter :)

**Reply by lbargaoanu** [reply]
> Thought that FooText would be mapped in both cases, because the <Bar, IFoo> map (which maps the property) is more derived than the <object, IFoo> map (which ignores the property).

It's more derived because of how you use `Include`. You're supposed to use `Include` between the closest levels in the hierarchy, not randomly. Just because `IncludeAllDerived` exists, it doesn't make it a good idea :)

**Reply by danielgreen** [reply]
> > Thought that FooText would be mapped in both cases, because the <Bar, IFoo> map (which maps the property) is more derived than the <object, IFoo> map (which ignores the property).
> 
> It's more derived because of how you use `Include`. You're supposed to use `Include` between the closest levels in the hierarchy, not randomly. Just because `IncludeAllDerived` exists, it doesn't make it a good idea :)

I must be too enthusiastic...

**Reply by lbargaoanu** [reply]
https://github.com/AutoMapper/AutoMapper/blob/95620ab0b250921a4b6aac63e309e18e94cb0bfc/src/UnitTests/MappingInheritance/IgnoreShouldBeInherited.cs#L113
On the other hand, I don't see why _that_ wouldn't work. And yet it seems it never did, so recent versions didn't change that. We'll see :)

**Reply by danielgreen** [reply]
Oh so one or more of those tests fail?
Sorry I tried to build the solution and run the tests myself to see, but hit a lot of package restore errors like this:
<confused face>

Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/flat2/microsoft.sourcelink.common/index.json'.
Response status code does not indicate success: 503 (Service Unavailable - TF10216: Azure DevOps services are currently unavailable. Try again later. Activity Id: f7450969-5311-4ce2-bff1-96a2f5616aaf (DevOps Activity ID: F7450969-5311-4CE2-BFF1-96A2F5616AAF)).
Enhanced retry: HttpRetryHandler is in a state that retry would have been abandoned or not waited if it were not enabled.
  ServiceUnavailable https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/flat2/microsoft.sourcelink.github/index.json 894ms

**Reply by lbargaoanu** [reply]
As far as I can tell, that test fails in recent AM versions. About the restore, I guess the cloud is busy today :)

**Reply by danielgreen** [reply]
Pulled the latest this morning, package restore was ok this time, and every unit test passes in the mapping_inheritance branch. I thought some of the tests in IgnoreShouldBeInherited.cs would fail but maybe I picked you up wrong?

**Reply by lbargaoanu** [reply]
Some did fail and I fixed them. We want to say that the new behaviour is that an ignore override is inherited. So, in your case, as you can see in the test, you only need to include the most derived map and you inherit both the config in the base class and the ignore override in the derived class.

**Reply by danielgreen** [reply]
The fix doesn't change the outcome when using the IncludeAllDerived approach that I started with - I assume that was not your intention anyway - I'm stepping through the source to try to understand why the execution plan is different when using IncludeAllDerived depending on the order of the maps in the profile...

**Reply by danielgreen** [reply]
Would you mind having a look at this commit on my fork please?
https://github.com/danielgreen/AutoMapper/commit/be66ea7c97588041cd4e4d415b294d4f46eab475

It's an attempt to sort the inherited TypeMaps before applying them, by looking at which inherited TypeMap is closest to the current TypeMap in the hierarchy. 

Since each TypeMap obviously has source and destination types with their own hierarchies, the TypeMapComparer works out the total number of hops from the current TypeMap's source & destination types to the inherited TypeMap source & destination types.
Those with the least hops are applied first, which means that the order the maps appear in the profile no longer has an effect when using IncludeAllDerived.

A caveat is where the TypeMaps use an interface as source or destination type. Since an interface does not appear in a type's hierarchy, TypeMapComparer treats them as having an arbitrary high number of hops in order to push them to the end of the sorted list. In other words, inherited maps that work on concrete classes will be applied before those that work on interfaces. For the purely interface-based maps, the order they appear in the profile will still matter.

The tests I've added (and all the other existing tests) pass.

**Reply by lbargaoanu** [reply]
I'm sorry, but that's exactly the kind of code I, for one, don't want to see in AM anymore. Complex and slow. And what exactly is the benefit here? To allow you to call `IncludeAllDerived`, which on every call would go through _all_ your maps, looking for derived maps.
I'd rather you implement smth like `IncludeAllDerived` in your own code somehow. Stuff is usually public so I think you can do it, but if smth is missing, we can expose it.

**Reply by danielgreen** [reply]
There would be a performance hit due to the extra sorting code so understandable if it's not worth it.

`IncludeAllDerived` already goes through all the maps doesn't it(?) to add the base maps into the internal inherited maps list maintained by all the derived maps, so there is no change there.

It's later on, when the inherited maps are about to be applied, that the sorting code runs. But yeah there's still a cost to running the sort.
